examples/non-fungible/src/service.rs

```diff
@@ -0,0 +1,236 @@
+};
+use thiserror::Error;
+
+linera_sdk::service!(NonFungibleToken);
+
+impl WithServiceAbi for NonFungibleToken {
+    type Abi = non_fungible::NonFungibleTokenAbi;
+}
+
+#[async_trait]
+impl Service for NonFungibleToken {
+    type Error = Error;
+    type Storage = ViewStateStorage<Self>;
+
+    async fn handle_query(
+        self: Arc<Self>,
+        _runtime: &ServiceRuntime,
+        request: Request,
+    ) -> Result<Response, Self::Error> {
+        let schema = Schema::build(
+            QueryRoot {
+                non_fungible_token: self.clone(),
+            },
+            MutationRoot,
+            EmptySubscription,
+        )
+        .finish();
+        let response = schema.execute(request).await;
+        Ok(response)
+    }
+}
+
+struct QueryRoot {
+    non_fungible_token: Arc<NonFungibleToken>,
+}
+
+#[Object]
+impl QueryRoot {
+    async fn nft(&self, token_id: String) -> Option<NftOutput> {
+        use base64::engine::{general_purpose::STANDARD_NO_PAD, Engine as _};
+        let token_id_vec = STANDARD_NO_PAD.decode(&token_id).unwrap();
+        let nft = self
+            .non_fungible_token
+            .nfts
+            .get(&TokenId { id: token_id_vec })
+            .await
+            .unwrap();
+
+        if let Some(nft) = nft {
+            Some(NftOutput {
+                token_id,
+                owner: nft.owner,
+                name: nft.name,
+                minter: nft.minter,
+                payload: nft.payload,
+            })
+        } else {
+            None
+        }
+    }
+
+    async fn nfts(&self) -> BTreeMap<String, NftOutput> {
+        use base64::engine::{general_purpose::STANDARD_NO_PAD, Engine as _};
+
+        let mut nfts = BTreeMap::new();
+        self.non_fungible_token
+            .nfts
+            .for_each_index_value(|token_id, nft| {
+                let token_id_str = STANDARD_NO_PAD.encode(token_id.id);
+                nfts.insert(
+                    token_id_str.clone(),
+                    NftOutput {
+                        token_id: token_id_str,
+                        owner: nft.owner,
+                        name: nft.name,
+                        minter: nft.minter,
+                        payload: nft.payload,
+                    },
+                );
+
+                Ok(())
+            })
+            .await
+            .unwrap();
+
+        nfts
+    }
+
+    async fn owned_token_ids_by_owner(&self, owner: AccountOwner) -> BTreeSet<String> {
+        use base64::engine::{general_purpose::STANDARD_NO_PAD, Engine as _};
+
+        let mut new_token_ids = BTreeSet::new();
+        let token_ids = self
+            .non_fungible_token
+            .owned_token_ids
+            .get(&owner)
+            .await
+            .unwrap();
+        for token_id in token_ids.into_iter().flatten() {
+            let new_token_id = STANDARD_NO_PAD.encode(token_id.id);
+            new_token_ids.insert(new_token_id);
+        }
+
+        new_token_ids
```

Comment:
Suggestion: use iterator composition functions here

```suggestion

        token_ids
            .into_iter()
            .flatten()
            .map(|token_id| STANDARD_NO_PAD.encode(token_id.id))
            .collect()
```

Reply:
Nice!